### ISSUE TITLE
Special case Chunk constructors for bit chunks

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3383]
Currently, a user can create a `Chunk[Boolean]`, which is backed up by an `Array[Boolean]`. However, booleans are not a very efficient way to represent bit data.

In order to improve efficiency, we should special case the construction of `Chunk[Boolean]`, and make sure that we always back it by `Chunk.BitChunk`, which is a special chunk subtype that encodes booleans using the bits inside a byte array (for a significant reduction in space usage, and corresponding increase in performance for bitwise operations).

A good first start would be looking at `Chunk.apply`, which is used in the chunk literal construction syntax: `Chunk(true, false, false, true)`, and ensuring that is represented by `Chunk.BitChunk`. But there are a number of other places where chunks are created and populated with data.

To deal with these other places, it will be convenient to define a `BitChunkBuilder`:

```scala
import scala.collection.mutable.Builder

trait BitChunkBuilder extends Builder[Boolean, BitChunk] {
  ...
}
```

A bit chunk builder will internally use an `ArrayBuilder` (defined in Scala collections) to build up an array of bytes: note that you have to add 8 bits into a bit chunk builder before you consume a full byte in the byte array.

Once the bit chunk builder is done, and the build method is invoked, the bit chunk builder will return the new `BitChunk` defined by the used bit range (keeping in mind the last byte inside the array may not be fully utilized, if the total number of bits written is not evenly divisible by 8).

With such a bit chunk builder, it should be straightforward (if a little tedious) to ensure that anytime a `Chunk[Boolean]` is being constructed, we don't actually use an `Array[Boolean]`, but rather, utilize a `BitChunkBuilder` to perform the construction.

**Comment by whysoserious** [comment]
I can take this one. 
I am coming back to Scala after a few years break and already have a lot of fun learning ZIO.

**Comment by jdegoes** [comment]
@whysoserious Welcome back, and thanks for your work!

After thinking about this, I believe the correct path forward is to modify the existing `ChunkBuilder` so it can built bit chunks. This will be a superior path to introducing a new bit chunk builder.

**Comment by whysoserious** [comment]
I'll take a look and prepare a proposal. Stay tuned.

**Comment by jdegoes** [comment]
See also: #3394

**Comment by jdegoes** [comment]
@whysoserious Now we have #3400, which introduces a builder for `Boolean` but it is only backed up by `Array[Boolean]`. So if you could do your work in `ChunkBuilder.Boolean`, that would be perfect!

**Comment by whysoserious** [comment]
On it! I'll probably need a couple of days for that.

**Comment by whysoserious** [comment]
PR: https://github.com/zio/zio/pull/3419

Looks good. Although I am not sure if I have used proper data structures in the builder.

**Comment by adamgfraser** [comment]
Closed by #3419.
